src/main/java/stream/flarebot/flarebot/permissions/PerGuildPermissions.java

```diff
@@ -2,112 +2,120 @@
 
 import net.dv8tion.jda.core.Permission;
 import net.dv8tion.jda.core.entities.Member;
+import net.dv8tion.jda.core.entities.Role;
 import stream.flarebot.flarebot.FlareBot;
 import stream.flarebot.flarebot.commands.Command;
 
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 
 public class PerGuildPermissions {
 
     private final ConcurrentHashMap<String, Group> groups = new ConcurrentHashMap<>();
-    private final ConcurrentHashMap<String, User> users = new ConcurrentHashMap<>();
-    protected String id;
+    private final ConcurrentHashMap<Member, User> users = new ConcurrentHashMap<>();
 
-    protected PerGuildPermissions() {
-    }
-
-    public PerGuildPermissions(String id) {
-        this.id = id;
+    public PerGuildPermissions() {
         if (!hasGroup("Default")) {
             for (Command command : FlareBot.getInstance().getCommands()) {
                 if (command.isDefaultPermission()) {
-                    addPermission("Default", command.getPermission());
+                    Group defaults = new Group("Default");
+                    groups.put("Default", defaults);
+                    defaults.addPermission(command.getPermission());
                 }
             }
         }
     }
 
     public boolean hasPermission(Member user, String permission) {
         // So we can go into servers and figure out any issues they have.
+        updateRoles(user);
         if (isCreator(user.getUser()))
             return true;
         if (user.isOwner())
             return true;
         if (user.getPermissions().contains(Permission.ADMINISTRATOR))
             return true;
+        if(isContributor(user.getUser()))
+            return true;
         PermissionNode node = new PermissionNode(permission);
-        return getUser(user).getGroups().stream()
+        boolean userPerm = getUser(user).getGroups().stream()
```

Comment:
I'd make this entire thing like thr for loop I showed you because 1, it would be more efficient and 2, it would be a lot cleaner than having a stream and a for loop both with a Boolean.